### PR TITLE
Allows rotation and setting of initial camera coordinates.

### PR DESCRIPTION
**Pull request by jeffreypaul15**
The added methods 
```py
def set_camera_coordinates(self, coord)
```
Takes a SkyCoord and sets the camera's coordinates with the specified coordinate frame, should we also allow for a separate frame to be passed here instead of converting it to the default one?
```py
def rotate_camera(self, angle: u.deg = None)
```
Takes a Quantity and rotates the camera relative to that particular frame. i.e. the rotation for each frame is already defined so this just adds to that rotation value.

For example :
```py
camera_position = SkyCoord(0*u.deg, 0*u.deg, 8*const.R_sun, obstime=m.observer_coordinate.obstime, frame=frames.HeliographicStonyhurst)
plotter.set_camera_coordinates(camera_position)

# Rotate the camera by a given angle
plotter.rotate_camera(90*u.deg)
```
`HeliographicStonyhurst` has a default roll (plotter.camera.roll) value of `90` so `plotter.rotate_camera(90*u.deg)` would just add 90, thereby rotating by 180 and producing the image below (rotated 90 deg clockwise against the user's perspective, I'm not sure if we want to rotate from the user's perspective or against):
image

Pyvista already has a `plotter.camera.roll` to specify a particular angle if required and we can point to their docs or we can have a separate method that sets rotation to a particular value instead of adding to it.

### CODE REVIEW

File: sunkit_pyvista/plotter.py [hunk 1]
```diff
@@ -59,6 +61,33 @@ def _coords_to_xyz(self, coords):
...
+        Parameters
+        ----------
+        coords : `astropy.coordinates.SkyCoord`
+            Coordinates of the camera.
+        """
+        camera_position = self._coords_to_xyz(coord)
+        pos = tuple(camera_position[0])
+        self.plotter.camera.position = pos
+
+    def rotate_camera(self, angle: u.deg = None):
```
Comment (nabobalis) [inline sunkit_pyvista/plotter.py:77]: Will this need a u.quantity decorator to ensure the unit type?

File: sunkit_pyvista/plotter.py [hunk 2]
```diff
@@ -59,6 +61,33 @@ def _coords_to_xyz(self, coords):
...
+        """
+        Rotates the camera by the specified value in degrees.
+
+        Parameters
+        ----------
+        angle : `astropy.units.Quantity`
+            The angle of rotation.
+        """
+        print(self.plotter.camera.roll)
```
Comment (nabobalis) [inline sunkit_pyvista/plotter.py:86]: I assume these were used for debugging?

**Comment by nabobalis** [comment]
> Takes a SkyCoord and sets the camera's coordinates with the specified coordinate frame, should we also allow for a separate frame to be passed here instead of converting it to the default one?

I don't understand, you mean to allow them bypass the frame transform?

> `HeliographicStonyhurst` has a default roll (plotter.camera.roll) value of `90` so `plotter.rotate_camera(90*u.deg)` would just add 90, thereby rotating by 180 and producing the image below (rotated 90 deg clockwise against the user's perspective

Where does the default roll come from? 

>I'm not sure if we want to rotate from the user's perspective or against):

You mean a negative rotate? 

> Pyvista already has a `plotter.camera.roll` to specify a particular angle if required and we can point to their docs or we can have a separate method that sets rotation to a particular value instead of adding to it.

I guess if we wrap it, it allows us to make sure units are passed into and get handled correctly?

File: sunkit_pyvista/plotter.py (hunk 2 above)
Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:86]: Ah yes, sorry

File: sunkit_pyvista/plotter.py (hunk 1 above)
Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:77]: Oh yeah I suppose.

**Comment by jeffreypaul15** [comment]
> I don't understand, you mean to allow them bypass the frame transform?

Yeah? Wouldn't that be required if a user wants to place the camera at an arbitrary point?

> Where does the default roll come from?

By default Pyvista sets it to -120. This is altered when the camera position is set.
Looking into this a little deeper, I found that setting `self.plotter.set_focus(mesh.center)` locks the position of the rotation and doesn't depend on the camera variable set. This is what should've been done initially.

> You mean a negative rotate?

Yeah, for now 0 deg (for any frame) is : 
image

rotating it by 90 deg would mean clockwise from the camera's perspective? Or against it?
Basically would we want 90 deg to produce : 
image 
or 
image

> I guess if we wrap it, it allows us to make sure units are passed into and get handled correctly?

Yeah I suppose

**Comment by nabobalis** [comment]
> > I don't understand, you mean to allow them bypass the frame transform?
> 
> Yeah? Wouldn't that be required if a user wants to place the camera at an arbitrary point?

It sounds like would be useful to me, I guess a keyword arg to bypass the transform?

> By default Pyvista sets it to -120. This is altered when the camera position is set.
> Looking into this a little deeper, I found that setting `self.plotter.set_focus(mesh.center)` locks the position of the rotation and doesn't depend on the camera variable set. This is what should've been done initially.

So we might need to set this value based on the input frame type?

> > You mean a negative rotate?
> 
> Yeah, for now 0 deg (for any frame) is :
> rotating it by 90 deg would mean clockwise from the camera's perspective? Or against it?
> Basically would we want 90 deg to produce :

Hmm, I would suggest clockwise if it positive, anticlockwise if it was negative. 
David probably have more insights here.

A side question, won't rotate need to take a camera axis argument?

**Comment by jeffreypaul15** [comment]
> So we might need to set this value based on the input frame type?

If rotation is frame dependent, then I'll have to look into this more with some help from David.

> A side question, won't rotate need to take a camera axis argument?

Yeah, I suppose that would make more sense.

**Comment by jeffreypaul15** [comment]
Okay, so after playing around with for a bit.
The default map is plotted in the `yz` axes where y is `lon` and z is `lat`.
This isn't really altered by pyvista, the roll value is set to -120 by default to capture all 3 axes when displaying any object but when the `focus` is set to a partuclar mesh ( the map in our case ), the focus is set in the `XY` plane. So inherently it gets plotted with the the camera focusing on the `XY` plane which looks something like : 

image
Where the focus is at the center of the map mesh.

Now, when setting the camera coordinates to
```py
camera_position = SkyCoord(0*u.deg, 0*u.deg, 8*const.R_sun, obstime=m.observer_coordinate.obstime, frame=frames.HeliographicStonyhurst)
```
the `roll` value stays the same because the camera isn't rotating about the `XY` plane, it's position stays the same but produces : 

image

Now the question is whether we leave this as the inital plot or straighten it by setting roll to +90 which produces : 

image

But has a default value of +90 and we'll have to mention this before plotting.

In the case of : 
```py
camera_position = SkyCoord(0*u.deg, 0*u.deg, 8*const.R_sun, obstime=m.observer_coordinate.obstime, frame=frames.HeliocentricInertial)
```
The plot produced the image below, because of the coordinates being behind the map.

image

To straighten this plot, we'll have to rotate the camera by -90 deg which is basically dependent on which frame we choose the `camera_coordinates` to be.

File: examples/skip_plot_map_in_3d.py [hunk 3]
```diff
@@ -40,4 +38,11 @@
                 frame='heliocentricinertial')
 plotter.plot_line(line)
 
-plotter.show(cpos=(-100, 0, 0))
+# Define a SkyCoord for to set the positon of the camera
+camera_position = SkyCoord(0*u.deg, 0*u.deg, 8*const.R_sun, obstime=m.observer_coordinate.obstime, frame=frames.HeliographicStonyhurst)
```
Comment (dstansby) [inline examples/skip_plot_map_in_3d.py:42]: Could we change this to `m.observer_coordinate` instead? It's simpler, and I think a common use case.

File: sunkit_pyvista/plotter.py [hunk 4]
```diff
@@ -29,7 +30,8 @@ def __init__(self, coordinate_frame=None):
         if coordinate_frame is None:
             coordinate_frame = HeliocentricInertial()
         self._coordinate_frame = coordinate_frame
-        self._plotter = pv.Plotter(off_screen=True)
+        self._plotter = pv.Plotter()
+        self._plotter.show_axes()
```
Comment (dstansby) [inline sunkit_pyvista/plotter.py:34]: What's the motivation behind this? I think it might be better to not show the axes, but open to reasons why we should keep them.

File: sunkit_pyvista/plotter.py [hunk 5]
```diff
@@ -59,6 +61,31 @@ def _coords_to_xyz(self, coords):
...
+        Parameters
+        ----------
+        coords : `astropy.coordinates.SkyCoord`
+            Coordinates of the camera.
+        """
+        camera_position = self._coords_to_xyz(coord)
+        pos = tuple(camera_position[0])
+        self.plotter.camera.position = pos
+
+    def rotate_camera(self, angle: u.deg = None):
```
Comment (dstansby) [inline sunkit_pyvista/plotter.py:77]: I think it would be better to make this a method that sets the camera rotation, instead of adding (or subtracting) a given amount of rotation. This way the rotation is always relative to the solar north pole when specified. Thoughts?

File: sunkit_pyvista/plotter.py [hunk 6]
```diff
@@ -97,6 +124,7 @@ def plot_map(self, m, **kwargs):
         cmap = kwargs.pop('cmap', m.cmap)
         mesh = self._pyvista_mesh(m)
         self.plotter.add_mesh(mesh, cmap=cmap, **kwargs)
+        self._plotter.camera_position = 'xz'
```
Comment (dstansby) [inline sunkit_pyvista/plotter.py:127]: What's this for?

**Review by dstansby — COMMENTED** [review]
Looking good 👍 - I've left a few questions as comments.

File: sunkit_pyvista/plotter.py (hunk 4 above)
Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:34]: I was using it for displaying the plots and trying to figure out how the rotation works, we can get rid of them

File: sunkit_pyvista/plotter.py (hunk 5 above)
Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:77]: Yes, that was what I initially thought of, but looking the comment I made above, the default map produces a plot like :
image

File: sunkit_pyvista/plotter.py (hunk 6 above)
Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:127]: This doesn't do anything, I needed some input on how we position the camera first so I left this in for now

Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:127]: I'm still a bit confused on how we set the inital plot, whether we leave it in it's rotated position or forcefully set `roll` to a particular value to make the plot upright?

Comment (dstansby) [reply sunkit_pyvista/plotter.py:127]: I think we should set `roll` to the correct value to make sure the solar north pole is pointing up.

File: examples/skip_plot_map_in_3d.py (hunk 3 above)
Comment (jeffreypaul15) [reply examples/skip_plot_map_in_3d.py:42]: image
That produces a plot really far away, is this the right behaviour?

Comment (dstansby) [reply examples/skip_plot_map_in_3d.py:42]: Yes, this looks right, since the default view angle (the width of the camera) is 30 degrees - https://docs.pyvista.org/core/camera.html#pyvista.Camera.view_angle. We should add a `view_angle` property to our `Plotter` that accepts a `u.deg` quantity as input, and sets the `view_angle` property on the pyvista plotter. Then it can be set to a smaller value to "zoom in" to the image. But we can do all this in the next PR.

File: sunkit_pyvista/plotter.py (hunk 5 above)
Comment (dstansby) [reply sunkit_pyvista/plotter.py:76]: I think what we should do here is
- Specify that the input to our `rotate_camera` method is relative to the solar north pole.
- Internally subtract 90 degrees from the value given to set the `plotter.camera.roll` property correctly
- Call this method as `self.rotate_camer(0*u.deg)` in the plotter `__init__` to set the default starting camera position to solar north up

I think we should also rename `rotate_camera` to `set_camera_rotation`.

Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:76]: Sure, that's what I assumed we'd do but the issue with that is the fact that different coordinate frames have different initial value set right? So, if we're rotating relative to the solar north pole, we'd just have to add to the current rotation because it's always pointing up anyway, thoughts?

Comment (dstansby) [reply sunkit_pyvista/plotter.py:76]: Ah, I forgot that the plotter could have different coordinate systems set. In that case the rotation should be relative to the z-axis of the coordinate system, which for the default coordinate frame is the solar north pole. I don't think this means that we want to be adding to the rotation each time though.

Comment (jeffreypaul15) [reply sunkit_pyvista/plotter.py:76]: I'll play around with this a bit tonight and see what the results are. I didn't find rotation wrt a single axes in the docs. I'll update this PR in a bit, thanks for the help :)

File: examples/skip_plot_map_in_3d.py (hunk 3 above)
Comment (jeffreypaul15) [reply examples/skip_plot_map_in_3d.py:42]: Ah, okay. I'll adjust that later then

**Review by dstansby — APPROVED** [review]

**Comment by dstansby** [comment]
I think this is good enough for now. I'll have a play at some point, and might come back with more suggestions, but lets merge 🎉
